## General Transportation Problem!


### Formulate LP

In [1]:
import pandas as pd
import numpy as np
import pulp
from pulp import *

In [2]:
# List of facilities 

facilities = ['Troy', 'Newark', 'Harrisburg']

# Dictionary for capacity of facilities

capacity = {'Troy' : 350 , 'Newark' : 775, 'Harrisburg' : 300}

# All retail stores 

retail_stores = ['Pittsburg', 'Cleveland', 'Buffalo', 'Philadelphia', 'Boston', 'New York', 'Providence', 'Hartford']

#Dictionayr of demand

demand = {'Pittsburg' : 200, 'Cleveland' : 100, 'Buffalo' : 125, 'Philadelphia' : 250, 'Boston' : 225, 
          'New York' : 400, 'Providence' : 50, 'Hartford' : 75}

# List of all costs 

costs = [[1.07, 1.02, 0.63, 0.50, 0.37, 0.33, 0.36, 0.25], [0.77, 1.06, 0.61, 0.19, 0.48, 0.03, 0.41, 0.27], [0.43, 0.70, 
        0.63, 0.23, 0.82, 0.36, 0.74, 0.61]]

# Converting cost data into dictionary 

costs_dict = makeDict([facilities, retail_stores], costs, 0)

In [3]:
costs_dict

defaultdict(<function pulp.utilities.__makeDict.<locals>.<lambda>()>,
            {'Troy': defaultdict(<function pulp.utilities.__makeDict.<locals>.<lambda>()>,
                         {'Pittsburg': 1.07,
                          'Cleveland': 1.02,
                          'Buffalo': 0.63,
                          'Philadelphia': 0.5,
                          'Boston': 0.37,
                          'New York': 0.33,
                          'Providence': 0.36,
                          'Hartford': 0.25}),
             'Newark': defaultdict(<function pulp.utilities.__makeDict.<locals>.<lambda>()>,
                         {'Pittsburg': 0.77,
                          'Cleveland': 1.06,
                          'Buffalo': 0.61,
                          'Philadelphia': 0.19,
                          'Boston': 0.48,
                          'New York': 0.03,
                          'Providence': 0.41,
                          'Hartford': 0.27}),
             'Harrisburg': de

### Initialize LP Model

In [4]:
# Creating an LP problem 

prob = LpProblem("General Transportation Problem", LpMinimize)

C:\ProgramData\Anaconda3\lib\site-packages\pulp\pulp.py:1352: UserWarning: Spaces are not permitted in the name. Converted to '_'
  warnings.warn("Spaces are not permitted in the name. Converted to '_'")


### Define Decision Variables

In [5]:
# Creating list of tuples for all possible routes between facilities and retail_stores 

routes = [(f,r) for f in facilities for r in retail_stores]

# A dictionary vars that contain the variable routes

vars = LpVariable.dicts("Route", (facilities, retail_stores), 0, None, LpInteger)

In [6]:
routes

[('Troy', 'Pittsburg'),
 ('Troy', 'Cleveland'),
 ('Troy', 'Buffalo'),
 ('Troy', 'Philadelphia'),
 ('Troy', 'Boston'),
 ('Troy', 'New York'),
 ('Troy', 'Providence'),
 ('Troy', 'Hartford'),
 ('Newark', 'Pittsburg'),
 ('Newark', 'Cleveland'),
 ('Newark', 'Buffalo'),
 ('Newark', 'Philadelphia'),
 ('Newark', 'Boston'),
 ('Newark', 'New York'),
 ('Newark', 'Providence'),
 ('Newark', 'Hartford'),
 ('Harrisburg', 'Pittsburg'),
 ('Harrisburg', 'Cleveland'),
 ('Harrisburg', 'Buffalo'),
 ('Harrisburg', 'Philadelphia'),
 ('Harrisburg', 'Boston'),
 ('Harrisburg', 'New York'),
 ('Harrisburg', 'Providence'),
 ('Harrisburg', 'Hartford')]

### Define Objective Function

In [7]:
# Adding the minimum objective function to 'prob'

prob += ( lpSum([vars[f][r] * costs_dict[f][r] for (f,r) in routes]), "Sum_of_Transportation_Costs")

### Define the Constraints 

In [8]:
# The supply maximum constriant is added to each node (facilities)

for f in facilities:
    prob += ( lpSum([vars[f][r] for r in retail_stores]) <= capacity[f], 
             "Sum_into_facilities_%s" % f,
    )
    
# The demand minimum constraint is addded to prob for each demand node (retail_store)

for r in retail_stores:
    prob += (
        lpSum([vars[f][r] for f in facilities]) >= demand[r],
        "Sum_into_retail_stores_%s" % r,
    )   

### Solve the model

In [9]:
# The problem is using Pulp's choice of solver

prob.solve()

1

In [10]:
# Print the variables's optimized value

for v in prob.variables():
    print(v.name, '=', v.varValue)

# The optimized objective function value is printed on the screen 

print("Value objective function =" , value(prob.objective))

Route_Harrisburg_Boston = 0.0
Route_Harrisburg_Buffalo = 0.0
Route_Harrisburg_Cleveland = 100.0
Route_Harrisburg_Hartford = 0.0
Route_Harrisburg_New_York = 0.0
Route_Harrisburg_Philadelphia = 0.0
Route_Harrisburg_Pittsburg = 200.0
Route_Harrisburg_Providence = 0.0
Route_Newark_Boston = 0.0
Route_Newark_Buffalo = 125.0
Route_Newark_Cleveland = 0.0
Route_Newark_Hartford = 0.0
Route_Newark_New_York = 400.0
Route_Newark_Philadelphia = 250.0
Route_Newark_Pittsburg = 0.0
Route_Newark_Providence = 0.0
Route_Troy_Boston = 225.0
Route_Troy_Buffalo = 0.0
Route_Troy_Cleveland = 0.0
Route_Troy_Hartford = 75.0
Route_Troy_New_York = 0.0
Route_Troy_Philadelphia = 0.0
Route_Troy_Pittsburg = 0.0
Route_Troy_Providence = 50.0
Value objective function = 411.75


### The total cost(imbalanced) for the month of January is $411.75

In [ ]:
# List of facilities 

facilities = ['Troy', 'Newark', 'Harrisburg']

# Dictionary for capacity of facilities

capacity = {'Troy' : 475 , 'Newark' : 475, 'Harrisburg' : 475}

# All retail stores 

retail_stores = ['Pittsburg', 'Cleveland', 'Buffalo', 'Philadelphia', 'Boston', 'New York', 'Providence', 'Hartford']

#Dictionayr of demand

demand = {'Pittsburg' : 200, 'Cleveland' : 100, 'Buffalo' : 125, 'Philadelphia' : 250, 'Boston' : 225, 
          'New York' : 400, 'Providence' : 50, 'Hartford' : 75}

# List of all costs 

costs = [[1.07, 1.02, 0.63, 0.50, 0.37, 0.33, 0.36, 0.25], [0.77, 1.06, 0.61, 0.19, 0.48, 0.03, 0.41, 0.27], [0.43, 0.70, 
        0.63, 0.23, 0.82, 0.36, 0.74, 0.61]]

# Converting cost data into dictionary 

costs_dict = makeDict([facilities, retail_stores], costs, 0)

In [ ]:
prob = LpProblem("General Transportation Problem", LpMinimize)\

# Creating list of tuples for all possible routes between facilities and retail_stores 

routes = [(f,r) for f in facilities for r in retail_stores]

# A dictionary vars that contain the variable routes

vars = LpVariable.dicts("Route", (facilities, retail_stores), 0, None, LpInteger)

In [ ]:
# Adding the minimum objective function to 'prob'

prob += ( lpSum([vars[f][r] * costs_dict[f][r] for (f,r) in routes]), "Sum_of_Transportation_Costs")

# The supply maximum constriant is added to each node (facilities)

for f in facilities:
    prob += ( lpSum([vars[f][r] for r in retail_stores]) <= capacity[f], 
             "Sum_into_facilities_%s" % f,
    )
    
# The demand minimum constraint is addded to prob for each demand node (retail_store)

for r in retail_stores:
    prob += (
        lpSum([vars[f][r] for f in facilities]) >= demand[r],
        "Sum_into_retail_stores_%s" % r,
    ) 

In [ ]:
# The problem is using Pulp's choice of solver

prob.solve()

# Print the variables's optimized value

for v in prob.variables():
    print(v.name, '=', v.varValue)

# The optimized objective function value is printed on the screen 

print("Value objective function =" , value(prob.objective))

### The total cost (balanced) for the month of January is $421.25